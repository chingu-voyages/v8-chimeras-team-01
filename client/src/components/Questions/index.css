
#questions {
  width: 100%;
  max-width: 900px;
  min-height: calc(100vh - 250px);
  margin: 0 auto;
}

.answers-wrapper {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-gap: 10px;
    color: #444;
    margin: 1em;
}

.question__display {
  font-size: 1.2em;
  padding: 1em 0.5em;
  margin: 0;
  width: 85%;
  border-bottom: 3px solid var(--color-gray);
  margin: 0 auto;
}

.questions__title {
  align-self: flex-end;
  margin-right: 1em;
  margin-top: 0;
  font-size: 1.2em;
}

.question__timer {
    margin-top: 20px;
}

.box {
  background-color: #444;
  color: #fff;
  border-radius: 5px;
  padding: 20px;
  font-size: 150%;
  border: 4px solid #444;
}

.image-wrapper {
    margin: 1em;
}

.hide {
    display: none;
}

.show {
    display: block;
}

.highlight {
    background-color: yellow;
    color: black;
}

.selected {
  border: 4px solid var(--color-orange);
}

.no-clicks {
  pointer-events: none;
}

/***** Media Queries *****/

@media only screen and (min-width: 600px) {

  .question__display {
    width: 100%;
  }

}
