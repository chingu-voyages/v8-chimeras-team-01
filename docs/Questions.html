<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>Class: Questions | Class: Questions</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/bootstrap.min.css">
    <link type="text/css" rel="stylesheet" href="styles/prettify-jsdoc.css">
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/tui-doc.css">

    
</head>
<body>
<nav class="lnb" id="lnb">
    <div class="logo" style="width: 150px; height: 150px">
        
            <a href="https://github.com/nhnent/tui.jsdoc-template" rel="noopener noreferrer" target="_blank">
                <img src="https://image.flaticon.com/icons/svg/477/477169.svg" width="100%" height="100%">
            </a>
        
    </div>
    <div class="title">
        <h1><a href="index.html" class="link">Class: Questions</a></h1>
        
    </div>
    <div class="search-container" id="search-container">
        <input type="text" placeholder="Search">
        <ul></ul>
    </div>
    
        <ol class="lnb-tab">
            <li id="api-tab">
                <a href="#"><h4>Docs</h4></a>
            </li>
            <li id="examples-tab">
                <a href="#"><h4>Tutorials</h4></a>
            </li>
        </ol>
    
    <div class="lnb-api hidden"><h3>Classes</h3><ul><li><a href="Questions.html">Questions</a><button type="button" class="hidden toggle-subnav btn btn-link">  <span class="glyphicon glyphicon-plus"></span></button><div class="hidden" id="Questions_sub"><div class="member-type">Members</div><ul class="inner"><li><a href="Questions.html#state">state</a></li></ul><div class="member-type">Methods</div><ul class="inner"><li><a href="Questions.html#.sendAnswer-FunctionusedtosendcomputedanswertotheHost.">sendAnswer - Function used to send computed answer to the Host.</a></li><li><a href="Questions.html#.timer-Atimertocountdown.">timer - A timer to countdown.</a></li></ul></div></li></ul></div><div class="lnb-api hidden"><h3>Global</h3><ul><li><a href="global.html#%255BbuttonSwitch%255D">[buttonSwitch]</a></li><li><a href="global.html#%255BtoggleGames%255D">[toggleGames]</a></li><li><a href="global.html#%255BtoggleInfo%255D">[toggleInfo]</a></li><li><a href="global.html#%255BtoggleView%255D">[toggleView]</a></li><li><a href="global.html#%255BuseState(gameList)%255D">[useState(gameList)]</a></li><li><a href="global.html#%255BuseState(showGames)%255D">[useState(showGames)]</a></li><li><a href="global.html#%255BuseState(view)%255D">[useState(view)]</a></li></ul></div>
</nav>
<div id="resizer"></div>

<div class="main" id="main">
    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>Questions<span class="signature">()</span><span class="type-signature"></span></h2>
        
    
</header>

<article>
    
        <div class="container-overview">
        
            

<dt>
    
        <h4 class="name" id="Questions">
            
                new <span class="type-signature"></span>Questions<span class="signature">()</span><span class="type-signature"></span>
                
            
            <div class="container-source members">
                <code><a href="Questions_Questions.js.html">Questions/Questions.js</a></code>,
                <code><a href="Questions_Questions.js.html#line5">line 5</a></code>
            </div>
            
        </h4>

        
    
</dt>
<dd>

    

    

    

    

    

    

    

    

    

    

    

    


<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>


    

</dd>

        
        </div>
    

    
        <h3 class="subsection-title">Extends</h3>

        


    <ul>
        <li>Component</li>
    </ul>


    

    

    

    

    

    

    
        <h3 class="subsection-title">Members</h3>

        <dl>
            

<dt>
<h4 class="name" id="state">
    <span class="type-signature"></span>state<span class="type-signature"></span>
    
    <div class="container-source members">
        <code><a href="Questions_Questions.js.html">Questions/Questions.js</a></code>,
        <code><a href="Questions_Questions.js.html#line26">line 26</a></code>
    </div>
    
</h4>


</dt>
<dd>
    

    <!--
    
    -->

    


    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>time</code></td>
            

            <td class="type">
            
                
<span class="param-type">Number</span>


            
            </td>

            

            

            <td class="description last"><p>A number to countdown from.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>question</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>Hold the current question.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>answers</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array</span>


            
            </td>

            

            

            <td class="description last"><p>An array of possible answers.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>correctAnswer</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>The correct answer to the question.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>chosenAnswer</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>Answer chosen by the player.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>username</code></td>
            

            <td class="type">
            
                
<span class="param-type">String</span>


            
            </td>

            

            

            <td class="description last"><p>Username of the player.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>totalQuestions</code></td>
            

            <td class="type">
            
                
<span class="param-type">Number</span>


            
            </td>

            

            

            <td class="description last"><p>Indicates how many questions there are in total.</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>counter</code></td>
            

            <td class="type">
            
                
<span class="param-type">Number</span>


            
            </td>

            

            

            <td class="description last"><p>Indicates which question we are currently at.</p></td>
        </tr>

    
    </tbody>
</table>




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>


    

    
</dd>

        </dl>
    

    
        <h3 class="subsection-title">Methods</h3>

        <dl>
                

<dt>
    
        <h4 class="name" id=".sendAnswer-FunctionusedtosendcomputedanswertotheHost.">
            
                <span class="type-signature"><span class="icon green">static</span> </span>sendAnswer - Function used to send computed answer to the Host.<span class="signature">()</span><span class="type-signature"></span>
                
            
            <div class="container-source members">
                <code><a href="Questions_Questions.js.html">Questions/Questions.js</a></code>,
                <code><a href="Questions_Questions.js.html#line38">line 38</a></code>
            </div>
            
        </h4>

        
    
</dt>
<dd>

    

    

    

    

    

    

    

    

    

    

    

    


<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>


    

</dd>

            
                

<dt>
    
        <h4 class="name" id=".timer-Atimertocountdown.">
            
                <span class="type-signature"><span class="icon green">static</span> </span>timer - A timer to countdown.<span class="signature">()</span><span class="type-signature"></span>
                
            
            <div class="container-source members">
                <code><a href="Questions_Questions.js.html">Questions/Questions.js</a></code>,
                <code><a href="Questions_Questions.js.html#line65">line 65</a></code>
            </div>
            
        </h4>

        
    
</dt>
<dd>

    

    

    

    

    

    

    

    

    

    

    

    


<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>


    

</dd>

            </dl>
    

    

    
</article>

</section>




</div>

<footer>
    <img class="logo" src="https://image.flaticon.com/icons/svg/477/477169.svg" style="width: 150px; height: 150px">
    <div class="footer-text">My awesome footer text</div>
</footer>
<script>prettyPrint();</script>
<script src="scripts/jquery.min.js"></script>
<script src="scripts/tui-doc.js"></script>
<script src="scripts/linenumber.js"></script>

    <script>
        var id = 'Questions_sub'.replace(/"/g, '_');
        var selectedApi = document.getElementById(id); // do not use jquery selector
        var $selectedApi = $(selectedApi);

        $selectedApi.removeClass('hidden');
        $selectedApi.parent().find('.glyphicon').removeClass('glyphicon-plus').addClass('glyphicon-minus');
        showLnbApi();
    </script>

</body>
</html>