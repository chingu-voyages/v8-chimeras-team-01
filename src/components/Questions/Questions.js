import React, { useEffect, useState } from 'react';
import Answer from './Answer';
import './index.css';
import Loader from '../Loader';

export default function Questions({ updateMyScore, myScore, sendAnswer, question, onQ, totalQ, handleIncrementQ, pushLocation, playersUpdated, updateHost, readyLeaderBoard, goLeaderboard }) {


  /**
   * [useEffect Hook]
   * @method  timeInterval
   * @description  [Hook to start/stop timer on mount/unmount]
   * @return {function} [Cleans up on un-mount by clearing interval]
   */
  useEffect(() => {
    const timeInterval = setInterval(() => timer(), 1000)
    return () => {
      clearInterval(timeInterval);
    }
  })

  const [localScore, setLocalScore] = useState(myScore);
  const [correct, setCorrect] = useState('wrong');
  const [answer, setAnswer] = useState('');


  const handleChoice = (event, correct, answer) => {
    preventClicks(event);
    setCorrect(correct);
    setAnswer(answer);
    let updatedScore = 0
    if (correct === "correct") {
      updatedScore = localScore + (5 * time);
    } else {
      updatedScore = localScore;
    }
    setLocalScore(updatedScore);
    updateMyScore(updatedScore)
  }

  /**
   * [time Hook]
   * @method time
   * @description [Hook to set initial time in timer and method to update that count]
   */
  const [time, setTime] = useState(10);


  /**
   * [isLastQ]
   * @param  [Holds boolean, true if this is the last quesiton in the array, false otherwise]
   * @return {Boolean}
   */
  const isLastQ = (onQ === totalQ) ? true : false;

  /**
   * @function highlightCorrect [Highlights Answer marked as correct]
   */
  const highlightCorrect = () => {
    let correct = document.querySelector('.correct');
    correct.classList.add('highlight');
  }


  /**
   * @function timer [Decrements the time in state and pushes to Leaderboard URL except last question, then pushes to Results URL]
   */
  const timer = () => {
    if (time === 0) {
      let newTime = (time - 1);
      setTime(newTime);

      // Highlighting the correct answer
      highlightCorrect();

      // send answers to host regardless
      sendAnswer(correct, answer, localScore)

      if (!isLastQ) {
        // pushLocation("/host/leaderboard");
      } else {
        // pushLocation("/host/results");
      }

    } else {
      let newTime = (time - 1);
      setTime(newTime);
    }

  }

  useEffect(() => {checkHost()})

  const checkHost = () => {
    if (playersUpdated === true) {
      updateHost();
    }
  }
  /* Once user clicks answer, prevent more answers from being chosen */
  /**
   * @function preventClicks [finds all Answers and prevents further clicks]
   */
  const preventClicks = (e) => {
    e.target.classList.add('selected');

    let allAnswers = document.getElementsByClassName('box');
    [...allAnswers].forEach(element => element.classList.add('no-clicks'));
  }

  return (

    <div id="questions">
      <h1>Question {onQ} of {totalQ} </h1>

      <div className='image-wrapper'>
        {time <= 0 ? <Loader /> :
          <>
            <img className='outline image-question' alt='of question' src='https://picsum.photos/200'></img>
            <div>Timer {time}</div>
          </>

        }
      </div>

      <div className='center'>
        <h1 className='inline'>{question.q}</h1>
      </div>

      <div className='answers-wrapper'>
        {
          question.a.map((answer, i) => {
            if (answer !== question.c) {
              return <Answer correct={'wrong'}
                answer={answer}
                handleChoice={handleChoice}
                key={i} />;
            } else {
              return <Answer correct={'correct'}
                answer={answer}
                handleChoice={handleChoice}
                key={i} />;
            }
          })
        }
      </div>

    </div >
  )
}
