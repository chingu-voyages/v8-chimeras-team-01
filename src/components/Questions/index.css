

.answers-wrapper {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-gap: 10px;
    color: #444;
    margin: 1rem;
}

.question__display {
  font-size: 1.2em;
  padding: 1em 0.2em;
  margin: 0;
  width: 100%;
  border-bottom: 3px solid var(--color-gray);
  margin: 0 auto;
}

.questions__title {
  align-self: flex-end;
  margin-right: 1em;
  font-size: 1.2em;
}

.question__timer {
    margin-top: 20px;
}

.box {
    background-color: #444;
    color: #fff;
    border-radius: 5px;
    padding: 20px;
    font-size: 150%;
    border: 4px solid #444;
}

.image-wrapper {
    margin: 1em;
}

.hide {
    display: none;
}

.show {
    display: block;
}

.highlight {
    background-color: yellow;
    color: black;
}

.selected {
  border: 4px solid var(--color-orange);
}

.no-clicks {
  pointer-events: none;
}
