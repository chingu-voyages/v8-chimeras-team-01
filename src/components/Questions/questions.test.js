import React from 'react';
import { shallow } from 'enzyme';

import Questions from './Questions.js';

describe('Questions', () => {

  it('should render correctly', () => {
    const component = shallow(<Questions />);
    expect(component).toMatchSnapshot();
  });

});
