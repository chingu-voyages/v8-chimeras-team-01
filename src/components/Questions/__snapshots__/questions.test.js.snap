// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions should render correctly 1`] = `
<div
  id="questions"
>
  <h1>
    Question 
    1
     of 
    10
     
  </h1>
  <div
    className="image-wrapper"
  >
    <img
      alt="of question"
      className="outline image-question"
      src="https://picsum.photos/200"
    />
    <img
      className="loader hide"
      src="./images/loader.gif"
    />
    <div>
      Timer 
      10
    </div>
  </div>
  <div
    className="center"
  >
    <h1
      className="inline"
    >
      The Question ?
    </h1>
  </div>
  <div
    className="answers-wrapper"
  >
    <Answer
      answer="Answer 1"
      correct="wrong"
      key="0"
      sendAnswer={[Function]}
    />
    <Answer
      answer="Answer 2"
      correct="correct"
      key="1"
      sendAnswer={[Function]}
    />
    <Answer
      answer="Answer 3"
      correct="wrong"
      key="2"
      sendAnswer={[Function]}
    />
    <Answer
      answer="Answer 4"
      correct="wrong"
      key="3"
      sendAnswer={[Function]}
    />
  </div>
</div>
`;
